File: src/main/java/pw/switchcraft/plethora/integration/MetaWrapper.java
package pw.switchcraft.plethora.integration;

import pw.switchcraft.plethora.api.meta.BaseMetaProvider;
import pw.switchcraft.plethora.api.method.IPartialContext;
import pw.switchcraft.plethora.api.reference.ConstantReference;

import javax.annotation.Nonnull;
import java.util.Map;
import java.util.Objects;

public class MetaWrapper<T> implements ConstantReference<MetaWrapper<T>> {
	private final T value;

	public MetaWrapper(@Nonnull T value) {
		Objects.requireNonNull(value, "value cannot be null");
		this.value = value;
	}

	@Nonnull
	public T value() {
		return value;
	}

	public static <T> MetaWrapper<T> of(@Nonnull T value) {
		return new MetaWrapper<>(value);
	}

	@Nonnull
	@Override
	public MetaWrapper<T> get() {
		return this;
	}

	@Nonnull
	@Override
	public MetaWrapper<T> safeGet() {
		return this;
	}

	public static final class MetaProvider extends BaseMetaProvider<MetaWrapper> {
		public MetaProvider() {
			super("Simply wraps an object and exposes metadata for that. You can happily ignore this.");
		}

		@Nonnull
		@Override
		public Map<String, ?> getMeta(@Nonnull IPartialContext<MetaWrapper> context) {
			return context.makePartialChild(context.getTarget().value()).getMeta();
		}
	}
}

File: src/main/java/io/sc3/plethora/integration/MetaWrapper.kt
package io.sc3.plethora.integration

import io.sc3.plethora.api.meta.BaseMetaProvider
import io.sc3.plethora.api.method.IPartialContext
import io.sc3.plethora.api.reference.ConstantReference

class MetaWrapper<T : Any>(val value: T) : ConstantReference<MetaWrapper<T>> {
  override fun get(): MetaWrapper<T> = this
  override fun safeGet(): MetaWrapper<T> = this

  object MetaProvider : BaseMetaProvider<MetaWrapper<*>>(
    description = "Simply wraps an object and exposes metadata for that. You can happily ignore this."
  ) {
    override fun getMeta(context: IPartialContext<MetaWrapper<*>>): Map<String, *> =
      context.makePartialChild(context.target.value).meta
  }

  companion object {
    @JvmStatic
    fun <T : Any> of(value: T): MetaWrapper<T> = MetaWrapper(value)
  }
}
